doc/advanced_data_structures.rst

```diff
@@ -0,0 +1,117 @@
+``aggregation list``, which can be referenced in all user-facing functions
+by the provided list name.  Note that a list can be appended by
+any client in the same or different application.  Implicitly, all
+DataSets, when appended, are added to the end of the list. If the list does not
+already exist, it is automatically created.
+
+Retrieving DataSets from an aggregation list
+--------------------------------------------
+
+To retrieve ``aggregation list`` contents,
+the SmartRedis ``Client`` method provides an API function that
+will return an iterable container with all of the ``DataSet`` objects
+appended to the ``aggregation list``.  For example, the C++ client
+function to retrieve the entire ``aggregation list`` contents is shown below:
+
+.. code-block:: cpp
+
+    # C++ aggregation list retrieval interface
+    std::vector<DataSet> get_datasets_from_list(const std::string& list_name);
+
+It is also possible to retrieve a subset of the DataSets within an aggregation
+list:
+
+.. code-block:: cpp
+
+    # C++ aggregation list subset retrieval interface
+    std::vector<DataSet> get_dataset_list_range(const std::string& list_name,
+                                                    const int start_index,
+                                                    const int end_index);
+
+An application can determine how many DataSets are in an aggregation List
+via the following API:
+
+.. code-block:: cpp
+
+    # C++ aggregation list length query
+    int get_list_length(const std::string& list_name);
+
+Synchronization
+---------------
+
+A SmartRedis Client can use the following APIs to block until a
+predicate is matched on the length of the list:
+
+.. code-block:: cpp
+
+    # Block until the list reaches a specific length
+    bool poll_list_length(const std::string& name, int list_length,
+                            int poll_frequency_ms, int num_tries);
+
+    # Block until the list reaches or exceeds a specific length
+    bool poll_list_length_gte(const std::string& name, int list_length,
+                                int poll_frequency_ms, int num_tries);
+
+    # Block until the list is no longer than a specific length
+    bool poll_list_length_lte(const std::string& name, int list_length,
+                                int poll_frequency_ms, int num_tries);
+
+Other operations
+----------------
+
+Finally, aggregation lists may be copied, renamed, or deleted. Note
+that there is no synchronization support for these operations; performing
+these operations when multiple applications are accessing the list may
+lead to race conditions:
+
+.. code-block:: cpp
+
+    # Copy an aggregation list
+    void copy_list(const std::string& src_name,
+                    const std::string& dest_name);
+
+    # Rename an aggregation list
+    void rename_list(const std::string& src_name,
+                        const std::string& dest_name);
+
+    # Delete an aggregation list
+    void delete_list(const std::string& list_name);
```

Comment:
(general comment, not a comment on this line)  I wonder if we should include a mention about the handling of aggregation lists within and across ensembles.

Reply:
Will open a separate ticket to make sure that any interactions between dataset aggregation and ensembles are covered. I think the only interaction should be with prefixing, but this needs to be researched.